core/domain/collection_domain.py

```diff
@@ -389,6 +453,56 @@ def _convert_collection_contents_v1_dict_to_v2_dict(
         return collection_contents
 
     @classmethod
+    def _convert_collection_contents_v2_dict_to_v3_dict(
+            cls, collection_contents):
+        """Converts from version 2 to 3.
+
+        Adds a skills dict and skill id counter. Gets skills in
+        prerequisite_skills and acquired_skills in nodes, and assigns them
+        integer IDs.
+        """
+        collection_contents['skills'] = {}
+        collection_contents['skills_id_count'] = 0
+
+        def _add_or_get_id(skill):
```

Comment:
+1 to set idea; it's more pythonic and probably more efficient. I think with this approach we'd also be able to create the skills dict at the end, rather than continuously. That might make ``_add_or_get_id`` small enough to just get rid of it (1 line if I'm thinking about this correctly).

Put another way I'm thinking of something like this:

```python
skill_names = list(set(all_prerequisite_skills + all_acquired_skills))
# For each collection node
acquired_skill_ids = [skill_names.index(acquired_skill_name) for acquired_skill_name in node['acquired_skills']] # ditto for prerequisite

skills = {
    'id': index,
    'name': skill_name,
    'question_ids': []
    for index, skill_name in enumerate(skill_names)
}
```